From 676e45179c18c2c27c5dfd05fee86a8b54b96183 Mon Sep 17 00:00:00 2001
From: Jose Dapena Paz <jdapena@igalia.com>
Date: Thu, 16 May 2024 13:44:10 +0200
Subject: Fix linking Rust libraries with GNU linkers

GNU linkers work as single-pass linkers, and unresolved symbols are
progressively resolved with later components being linked. That means
users of a library should be linked before the library.

As Rust libraries are sorted by name, linking will break. To solve
this, put all Rust libraries in a linking group.

Bug: 40945821
Change-Id: Ie064ec05264673ac9756953a764da5588923ed4b
---
 build/toolchain/gcc_toolchain.gni | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/build/toolchain/gcc_toolchain.gni b/build/toolchain/gcc_toolchain.gni
index b737774cac4a9..fd3ef37b42c4b 100644
--- a/build/toolchain/gcc_toolchain.gni
+++ b/build/toolchain/gcc_toolchain.gni
@@ -611,7 +611,10 @@ template("single_gcc_toolchain") {
         start_group_flag = "-Wl,--start-group"
         end_group_flag = "-Wl,--end-group "
       }
-      link_command = "$ld {{ldflags}}${extra_ldflags} -o \"$unstripped_outfile\" $start_group_flag @\"$rspfile\" {{solibs}} $end_group_flag {{libs}} {{rlibs}}"
+      # We need to specify link groups, at least, for single pass linkers. I.e.
+      # Rust libraries are alphanumerically shorted instead of ordered by
+      # dependencies so they fail to link if not properly ordered or grouped.
+      link_command = "$ld {{ldflags}}${extra_ldflags} -o \"$unstripped_outfile\" $start_group_flag @\"$rspfile\" $end_group_flag {{solibs}} {{libs}} $start_group_flag {{rlibs}} $end_group_flag"
 
       # Generate a map file to be used for binary size analysis.
       # Map file adds ~10% to the link time on a z620.
-- 
2.43.0

